### DISCUSSION TITLE
Analyze continuously acquired analog signal from Arduino

### DISCUSSION THREAD

**Discussion by mgraupe** [bonsai-rx/bonsai#1406] [Q&A] [answered]
Hello, 

For a closed-loop experiment, we aim to continuously sample the analog input of an Arduino, perform analysis on the signal and trigger a digital output based on this analysis. However, many transformation tools do not seems to work on the data stream.

We currently, sample A0 of the Arduino and can perform subtraction, conversion to absolute value, for example. However, we cannot implement  thresholding, summing or calculating the difference between values in the data stream (we are interested in the derivative). The given error message when connecting any of those elements is "No method overload found for the given arguments.". The operations work on an input stream from a camera but not from the Arduino. 

Do you have any idea what is required to make these computations possible on our Arduino data source (AnalogInput)? Thanks a lot for your help in advance. 

Best,
Michael 
![Screenshot 2023-06-08 170024](https://github.com/bonsai-rx/bonsai/assets/8803795/dcef301a-37aa-453c-8332-2b6ac3cd11b4)

**Comment by bruno-f-cruz** [comment]
When working on top of sequences of numbers you will probably want to use add or sum(reactive), depending on the intended outcome.

**Reply by mgraupe** [reply]
Dear Bruno, 

Thank you for your response. I have managed to do summing over the last 6 values in a sequence with the workflow below. Is there a more intelligent way to do this if I want to calculate the running sum over the last 100 elements of a sequence, for example? 

Thanks,
Michael
![Screenshot 2023-06-12 160137](https://github.com/bonsai-rx/bonsai/assets/8803795/f09c79b3-68ae-4053-ae35-41f4ad449149)

**Comment by bruno-f-cruz** [comment, marked as answer]
![image](https://github.com/bonsai-rx/bonsai/assets/7049351/57b400d8-992f-45c8-b69e-399ce3cc1d7a)
With `Count` being the size of your window and `Skip` determines the number of overlapping samples.

**Comment by mgraupe** [comment]
Thank you for the suggestion, it works well for integrating the signal.

We would like to perform the subsequent analysis of the sequence with PythonTransform. How can I access several values in the sequence, i.e., values from different time-points. Something like `value[-1], value[-2], value[-3]`, etc. ? Is that possible and if yes, how can that be implemented? 

Thanks in advance. 

Cheers,
Michael
